staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/validation_test.go

```diff
@@ -23,6 +23,80 @@ import (
 	fuzz "github.com/google/gofuzz"
 )
 
+func TestValidateStructuralInvariantsMetadata(t *testing.T) {
+	fuzzer := fuzz.New()
+	fuzzer.Funcs(
+		func(s *JSON, c fuzz.Continue) {
+			if c.RandBool() {
+				s.Object = float64(42.0)
+			}
+		},
+		func(s **StructuralOrBool, c fuzz.Continue) {
+			if c.RandBool() {
+				*s = &StructuralOrBool{}
+			}
+		},
+		func(s **Structural, c fuzz.Continue) {
+			if c.RandBool() {
+				*s = &Structural{}
+			}
+		},
+		func(s *Structural, c fuzz.Continue) {
+			if c.RandBool() {
+				*s = Structural{}
+			}
+		},
+		func(vv **ValueValidation, c fuzz.Continue) {
+			if c.RandBool() {
+				*vv = &ValueValidation{}
+			}
+		},
+	)
+	fuzzer.NilChance(0).NumElements(1,1)
+
+	// check that anything other than name and generateName in metadata properties is forbidden
+	properties := []string {"name", "generateName", "testProperty"}
+	for _, property := range properties {
+		s := &Structural{
+			Properties: map[string]Structural{
+				"metadata": {
+					Properties: map[string]Structural{
+						property: {},
+					},
+				},
+			},
+		}
+		errs := validateStructuralInvariants(s, rootLevel, nil)
+		if len(errs) != 0 {
+			t.Logf("errors returned: %v", errs)
+		}
+		if len(errs) == 0 && property != "name" && property != "generateName" {
+			t.Errorf("expected forbidden field validation errors for: %#v", s)
+		}
+	}
+
+	// check that anything other than type and properties in metadata is forbidden
+	tt := reflect.TypeOf(Structural{})
+	for i := 0; i < tt.NumField(); i++ {
+		metadata := Structural{}
+		x := reflect.ValueOf(&metadata).Elem()
+		fuzzer.Fuzz(x.Field(i).Addr().Interface())
+
+		s := &Structural{
+			Properties: map[string]Structural{
+				"metadata": metadata,
+			},
+		}
+		errs := validateStructuralInvariants(s, rootLevel, nil)
+		if len(errs) != 0 {
+			t.Logf("errors returned: %v", errs)
+		}
+		if len(errs) == 0 && metadata.Properties == nil && metadata.Type == "" {
```

Comment:
cannot follow this line. Why is `len(errs) == 0` good just because properties or type is non-nil/non-empty?